fix(doc): improve doc generation of types using lpeg

lewis6991 opened the pull request:
- Added an lpeg grammar for LuaCATS (with typing)
- Use it in `lua2dox.lua` and simplified it.
- All types are now rendered as ```(`type`)```
- Render optional types using `?` instead of `|nil`

lewis6991 committed:
fix(doc): improve doc generation of types using lpeg

Added a lpeg grammar for LuaCATS and use it in lua2dox.lua

wookayin commented on scripts/lua2dox.lua:
```diff
@@ -54,6 +54,7 @@ The effect is that you will get the function documented, but not with the parame
 ]]
 
 local TYPES = { 'integer', 'number', 'string', 'table', 'list', 'boolean', 'function' }
+local TYPE_PAT = '[a-zA-Z][a-zA-Z_.]*'
```
This is already a great improvement, but it seems that more complex types like `any[]|nil` are not captured.

```suggestion
local TYPE_PAT = '[a-zA-Z][a-zA-Z_.%[%]]*'
```

With the above we can deal with `foo[]|nil` types. Function types that contain `?`, `,` `(`, `)`, etc. might be quite complicated to process with regex (we'll need treesitter eventually....), so this might be addressed as a separate scope.

wookayin replied:
We can also capture the `optional?` types:

```suggestion
local TYPE_PAT = '[a-zA-Z][a-zA-Z_.%[%]|?]*'
```
There are only a few such occurrences, `(string?)`, `(Version ?|nil)`, etc.

lewis6991 replied:
I don't think treesitter would ever be the right tool for this. Instead I've implemented an lpeg grammar to capture many more types.

clason replied:
FWIW, (unlocking) improvements in our tooling is a valid argument for standardizing type annotations.

lewis6991 commented on runtime/doc/lua.txt:
```diff
@@ -860,7 +860,7 @@ vim.spell.check({str})                                     *vim.spell.check()*
 
     Return: ~
-        `{[1]: string, [2]: string, [3]: string}[]` List of tuples with three
+        ({[1]: string, [2]: string, [3]: string}[]) List of tuples with three
```
This is sort of a regression, however I think we should either surround all types with backticks or not at all.

@clason?

clason replied:
I think the backticks help with reading complex return types, but I agree: since the vast majority of type annotations uses parens, deviating adds (a little) friction. So they're not worth keeping around, especially if this makes docgen easier/more robust.

wookayin replied:
+1 for backticks. Parens are not highlighted.

If we use backticks for all types, some will lose links to helptags (e.g. |TSNode|), but I think having links would be useful for some types.

lewis6991 replied:
The backticks are trivial to add, it's really a matter of preference. I would vote to add them since it reduces the change of vimdoc parse errors as types can be complicated expressions. The downside would that we wouldn't be able to use the types as tags.

Don't need to make a decision now, I've not included them in this PR to reduce the diff.

clason replied:
Either is fine with me, I trust your opinion on which is more valuable. (The links can always be moved into the annotation comment, so we don't _necessarily_ lose them.)

lewis6991 replied:
Ok, we can hold off for now and see how the vimdoc parser copes.

clason replied:
I'm fine with adding backticks now, for the record.

wookayin replied:
For links: it seems that `CTRL-]` on the following anchor can work, correctly jumps to `*TSNode*`:

```
|`TSNode`|

:help |`TSNode`|
```

However, this doesn't work (while `CTRL-]` can jump):
```
:help `|TSNode|`
```

wookayin commented on test/functional/luacats_grammar_spec.lua:
```diff
@@ -0,0 +1,108 @@
+    desc = 'the command and arguments'
+  })
+
+  test('@return (string command, string[] args) the command and arguments', {
```
I think multiple (variable) returns is like: `@return string...\n@return string[]...`

https://luals.github.io/wiki/annotations/#return

lewis6991 replied:
It's both:

```lua
@return string a, string b
```

is equivalent to:
```
@return string a
@return string b
```

wookayin replied:
Oh didn't know that. TIL, thanks! I'm gonna fix the luadoc treesitter parser.

wookayin commented on test/functional/luacats_grammar_spec.lua:
```diff
@@ -0,0 +1,108 @@
+    desc = 'desc'
+  })
+
+  test('@return (string command) the command and arguments', {
```
is this a valid LuaCATS annotation for named return?

lewis6991 replied:
LuaLS supports it though it might not be documented.

wookayin reviewed (commented):
I love the new lpeg grammar 👍

lewis6991 commented:
- I've added backticks since there were some vimdoc errors, and tbh I think it makes complicated types easier to read.
- Converted `|nil` -> `?` since we can now say we are just using LuaCATS notation.

clason commented on runtime/doc/treesitter.txt:
```diff
@@ -630,75 +631,75 @@ get_node({opts})                                   *vim.treesitter.get_node()*
 
     Parameters: ~
-      • {node_or_range}  (|TSNode| | table) Node or table of positions
+      • {node_or_range}  (`TSNode|table`) Node or table of positions
```
💯 Working around those pipe conflicts was annoying.

lewis6991 replied:
I forgot about that. I guess that's another reason to not do the tagging since it conflict too much with unions.

clason replied:
Yeah. The loss of help links is a blow for treesitter specifically (I don't think we use much userdata elsewhere?), but that help needs more love anyway. 

We can always tack a simple `|TSNode|` at the end of the annotation comment (if there's not a link elsewhere in the docs already).

clason commented on runtime/doc/treesitter.txt:
```diff
@@ -940,8 +939,8 @@ lint({buf}, {opts})                              *vim.treesitter.query.lint()*
-      • {buf}   (integer) Buffer handle
-      • {opts}  (QueryLinterOpts|nil) Optional keyword arguments:
+      • {buf}   (`integer`) Buffer handle
+      • {opts}  (`table?`) Optional keyword arguments:
```
Loss of typing?

lewis6991 replied:
I did this on purpose since the type isn't used anywhere else.

Basically, if you do:
```
--- @param name some.complicated.type (table) desc
```

It will render the type as `table` instead of `some.complcated.type`.

clason replied:
Ah, but these aren't documented, so it's either weird type here or loss of typing in luals.

clason commented:
> Converted |nil -> ? since we can now say we are just using LuaCATS notation.

Very nice. Should we document this somewhere (for _users_, who may not read `develop.txt`)?

lewis6991 commented:
Yes I considered that, but didn't know where.

clason commented:
Yeah, that's the big question. Maybe in `lua.txt` somewhere? Or at the top of  `api.txt`?

Not critical, though, and can be left for later.

justinmk commented on runtime/doc/diagnostic.txt:
```diff
@@ -475,55 +475,55 @@ config({opts}, {namespace})                          *vim.diagnostic.config()*
                        ERROR is displayed before WARN). Options:
                        • reverse: (boolean) Reverse sort order
-      • {namespace}  (integer|nil) Update the options for the given namespace.
+      • {namespace}  (`integer?`) Update the options for the given namespace.
```
Did we consider dropping the parens? Especially since we have backticks now.

lewis6991 replied:
No but we could do. I kind of like the parens since it makes it clear it is separate from the description, but I don't feel super strong about it. The backticks will be concealed but will add highlighting.
